tornado-drivers/spirv/src/main/java/uk/ac/manchester/tornado/drivers/spirv/graal/phases/TornadoTaskSpecialization.java

```diff
@@ -294,8 +298,34 @@ private void propagateParameters(StructuredGraph graph, ParameterNode parameterN
                 parameterNode.replaceAtUsages(kernelContextAccessNode);
                 index++;
             } else {
-                ConstantNode constant = createConstantFromObject(args[parameterNode.index()], graph);
-                parameterNode.replaceAtUsages(constant);
+                var value = args[parameterNode.index()];
+                ConstantNode primitiveConstant = createPrimitiveConstantFromObjectParameter(value, graph);
+
+                parameterNode.replaceAtAllUsages(primitiveConstant, true);
+                parameterNode.safeDelete();
+
+                //remove Unbox nodes, they are not needed for constant values and cause compilation errors
+                graph.getNodes().filter(n -> n instanceof PiNode piNode && piNode.object() == primitiveConstant).snapshot().forEach(node -> {
```

Comment:
Basically, at this point, the original node could be surrounded by a `<Box/Unbox>Node` node and a `FixedGuardNode` with the check if `IsNullNode`, correct? If so, can you add a comment explaining this process?

Reply:
Yeah. Sure will do tomorrow. I wanted to add comment  but forgot about it :-) otherwise reasoning looks cryptic